### PR TITLE
N21-2014 Setup new SHD client

### PR DESCRIPTION
**Pull request by MarvinOehlerkingCap**
# Description

## Links to Tickets or other pull requests
- https://ticketsystem.dbildungscloud.de/browse/N21-2104
- https://github.com/hpi-schul-cloud/dof_app_deploy/pull/926

## Approval for review

- [ ] All points were discussed with the ticket creator, support-team or product owner. The code upholds all quality guidelines from the PR-template.

> Notice: Please remove the WIP label if the PR is ready to review, otherwise nobody will review it.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+          SONAR_TOKEN: ${{ secrets.SONARCLOUD_TOKEN }}
+        with:
+          args: >
+            -Dsonar.organization=schulcloud-verbund
+            -Dsonar.projectKey=hpi-schul-cloud_shd-client
+            -Dsonar.sources=.
+            -Dsonar.exclusions=src/serverApi/**/*.*
+            -Dsonar.coverage.exclusions=tests/**/*.*,**/*.unit.ts,**/*.unit.js
+            -Dsonar.cpd.exclusions=tests/**/*.*,**/*.unit.ts,**/*.unit.js,**/locales/*.ts
+            -Dsonar.javascript.lcov.reportPaths=./coverage/lcov.info
```
Comment (mamutmk5) [inline .github/workflows/test.yml:46]: No, please add an sonar-project.properties file to the project root and set this there.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          SONAR_TOKEN: ${{ secrets.SONARCLOUD_TOKEN }}
+        with:
+          args: >
+            -Dsonar.organization=schulcloud-verbund
+            -Dsonar.projectKey=hpi-schul-cloud_shd-client
+            -Dsonar.sources=.
+            -Dsonar.exclusions=src/serverApi/**/*.*
+            -Dsonar.coverage.exclusions=tests/**/*.*,**/*.unit.ts,**/*.unit.js
+            -Dsonar.cpd.exclusions=tests/**/*.*,**/*.unit.ts,**/*.unit.js,**/locales/*.ts
```
Comment (mamutmk5) [inline .github/workflows/test.yml:45]: To configure the test rigth set at the `sonar-project.properties` file then follow lines:
```
sonar.tests=.
sonar.test.inclusions=tests/**/*.*,**/*.unit.ts,**/*.unit.js
```
This says where sonar find the test and what folder and files are includet for test.
It also exclude the test from coverage, cpd and more code analyses

**Review by mamutmk5 — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# build stage
+FROM docker.io/node:18-bullseye AS build-stage
```
Comment (mamutmk5) [inline Dockerfile:2]: Use node 20 not more node 18

File: .dockerignore [hunk 4]
```diff
@@ -0,0 +1,11 @@
+.github
+.vscode
+deploy
+node_modules
+.travis.yml*
```
Comment (mamutmk5) [inline .dockerignore:5]: Not more need this 
```suggestion
```

File: .nvmrc [hunk 5]
```diff
@@ -0,0 +1 @@
+18
```
Comment (mamutmk5) [inline .nvmrc:1]: Node 20
```suggestion
20
```

File: .pullapprove.yml
Comment (mamutmk5) [inline .pullapprove.yml:1]: no pullapporve, we don't use it any more

**Review by mamutmk5 — CHANGES_REQUESTED** [review]

File: sonar-project.properties [hunk 6]
```diff
@@ -0,0 +1,14 @@
+sonar.organization=schulcloud-verbund
+sonar.projectKey=hpi-schul-cloud_shd-client
+
+sonar.sources=.
+sonar.tests=.
+
+# Exclude test files, locales and generated code from source scope
+sonar.exclusions=tests/**/*,**/*.unit.ts,**/*.unit.js,src/serverApi/**/*,**/locales/**.ts
```
Comment (mamutmk5) [inline sonar-project.properties:8]: The `tests/**/*` ,`**/*.unit.ts` and `**/*.unit.js` will be automaticly excluded as they add to `test.inclusions`

File: sonar-project.properties [hunk 7]
```diff
@@ -2,12 +2,12 @@ sonar.organization=schulcloud-verbund
 sonar.projectKey=hpi-schul-cloud_shd-client
 
 sonar.sources=.
-sonar.tests=.
+sonar.tests=src/
```
Comment (mamutmk5) [inline sonar-project.properties:5]: lett it be by `.`

File: openapitools-for-server.json [hunk 8]
```diff
@@ -0,0 +1,4 @@
+{
+	"supportsES6": true,
+	"withInterfaces": true
+}
```
Comment (SevenWaysDP) [inline openapitools-for-server.json:4]:
```suggestion
{
	"apiPackage": "api",
	"enumNameSuffix": "",
	"enumPropertyNaming": "UPPERCASE",
	"modelPackage": "models",
	"supportsES6": true,
	"withInterfaces": true,
	"withSeparateModelsAndApi": true
}
```

File: ansible/roles/shd-client-core/templates/ingress.yml.j2 [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    http:
+      paths:
+      ### SHD
+      - path: /
+        pathType: Prefix
+        backend:
+          service:
+            name: shd-client-svc
+            port:
+              number: {{ PORT }}
```
Comment (mamutmk5) [inline ansible/roles/shd-client-core/templates/ingress.yml.j2:40]: code the port hard

File: ansible/roles/shd-client-core/templates/svc.yml.j2 [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+kind: Service
+metadata:
+  name: shd-client-svc
+  namespace: {{ NAMESPACE }}
+  labels:
+    app: shd-client
+spec:
+  type: ClusterIP
+  ports:
+    - port: {{ PORT }}
```
Comment (mamutmk5) [inline ansible/roles/shd-client-core/templates/svc.yml.j2:11]: code the port hard

File: ansible/roles/shd-client-core/templates/deployment.yml.j2 [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+          httpGet:
+            path: /version
+            port: 4100
+          timeoutSeconds: 4
+          failureThreshold: 5
+          periodSeconds: 5
+        resources:
+          limits:
+            cpu: {{ SHD_CLIENT_CPU_LIMITS|default("1000m", true) }}
+            memory: {{ SHD_CLIENT_MEMORY_LIMITS|default("256Mi", true) }}
```
Comment (mamutmk5) [inline ansible/roles/shd-client-core/templates/deployment.yml.j2:85]: at least as Limit 2GB

File: ansible/roles/shd-client-core/templates/deployment.yml.j2 [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+          requests:
+            cpu: {{ SHD_CLIENT_CPU_REQUESTS|default("100m", true) }}
+            memory: {{ SHD_CLIENT_MEMORY_REQUESTS|default("32Mi", true) }}
```
Comment (mamutmk5) [inline ansible/roles/shd-client-core/templates/deployment.yml.j2:88]: 32 MB memoray are a littelbit to less, use 128MB

File: ansible/roles/shd-client-core/templates/configmap.yml.j2 [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: shd-client-configmap
+  namespace: {{ NAMESPACE }}
+  labels:
+    app: shd-client
+data:
+  # general
+  NODE_ENV: "production"
```
Comment (mamutmk5) [inline ansible/roles/shd-client-core/templates/configmap.yml.j2:11]: Where is it used?

**Review by mamutmk5 — CHANGES_REQUESTED** [review]

File: ansible/roles/shd-client-core/templates/configmap.yml.j2 (hunk 13 above)
Comment (MarvinOehlerkingCap) [reply ansible/roles/shd-client-core/templates/configmap.yml.j2:11]: It is used in the vue.config.js but I think that one is controlled by the Dockerfile during build, so I will remove it here

Comment (MarvinOehlerkingCap) [reply ansible/roles/shd-client-core/templates/configmap.yml.j2:11]: Do I need TZ? I have looked into the other ansibles and they all have NODE_ENV and TZ.

Comment (mamutmk5) [reply ansible/roles/shd-client-core/templates/configmap.yml.j2]: It was not about TZ it was about the NODE_ENV.
The reason why i ask if a veu/ts/js app who is put after build into an nginx really need NODE_ENV?

Comment (mamutmk5) [reply ansible/roles/shd-client-core/templates/configmap.yml.j2]: Top, has not seen that you removed it in your last commit :)

File: .github/workflows/push.yml [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+          else
+            echo "branch=${GITHUB_REF#refs/heads/}" >> $GITHUB_OUTPUT
+            echo "sha=${{ github.sha }}" >> $GITHUB_OUTPUT
+          fi
+
+  deploy:
+    needs:
+      - build_and_push
+      - branch_meta
+    uses: hpi-schul-cloud/dof_app_deploy/.github/workflows/deploy.yml@main
```
Comment (mamutmk5) [inline .github/workflows/push.yml:103]: As lang as you have not merge the dof_app_deploy an this pr to main use in this PR the branch as taget and set it back to main bevor you merge
```suggestion
    uses: hpi-schul-cloud/dof_app_deploy/.github/workflows/deploy.yml@N21-2104-setup
```

File: src/styles/settings.scss
Comment (uidp) [inline src/styles/settings.scss:1]: This looks a lot like a copy of `settings.scss`. I think it should be a file containing global styles + import of `settings.scss` + import of vuetify. See:
https://github.com/hpi-schul-cloud/nuxt-client/blob/3f97b312e95592923d9f61ddd406fd5132846867/src/styles/global.scss#L1-L23

**Review by mamutmk5 — APPROVED** [review]

**Review by uidp — APPROVED** [review]

**Pull request merged**